mreg/exceptions.py

```diff
@@ -0,0 +1,86 @@
+"""
+Custom exception handler for Django REST Framework (DRF).
+
+This module provides a custom exception handler that logs all exceptions
+with additional context for improved observability and debugging.
+"""
+
+import traceback
+
+import structlog
+from drf_standardized_errors.handler import ExceptionHandler
+from rest_framework import exceptions
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from mreg.middleware.logging_http import get_request_user_agent, get_request_username
+
+mreg_logger = structlog.getLogger("mreg.http")
+
+
+class MregExceptionHandler(ExceptionHandler):
+    """Custom DRF exception handler that logs all exceptions with additional context.
+
+    This handler extends the default DRF exception handler to log _all_ exceptions,
+    not just those that are handled by drf-standardized-errors (5xx errors).
+
+    Furthermore, it redacts sensitive information from `PermissionDenied` exceptions
+    to avoid leaking this information to users.
+    """
+
+    def report_exception(self, exc: exceptions.APIException, response: Response) -> None:
+        try:
+            drf_request: Request | None = self.context["request"]
+            request = drf_request._request  # pyright: ignore[reportOptionalMemberAccess, reportPrivateUsage]
+        except AttributeError:
+            # If the request is not available, we cannot log request-specific information
+            # NOTE: perform some sort of fallback logging here if needed
+            return super().report_exception(exc, response)
```

Comment:
I'm not sure how realistic this is. If we have a `Response` object, then _surely_ we have a `Request` as well? However, the type annotations for the context object don't guarantee the existence of a request, so here we are.